<launch>

    <arg name="map_path" default="/home/robot/htc-robot-ros_ws/data/map" />

    <!-- NDT Parameters -->
    <arg name="max_iter" default="30" doc="max iteration for alignment"/>
    <arg name="step_size" default="0.1" doc="step_size maximum step length[m]"/>
    <arg name="ndt_res" default="5.0" doc="resolution side length of voxels[m]"/>
    <arg name="trans_eps" default="0.01" doc="transform epsilon to stop iteration"/>
    <arg name="voxel_leaf_size" default="1.0" doc="a down sample size of a input cloud[m]"/>

    <!-- Mapping Parameters -->
    <arg name="scan_rate" default="10.0" doc="scan_rate"/>
    <arg name="min_scan_range" default="0" doc="min_scan_range"/>
    <arg name="max_scan_range" default="200.0" doc="max_scan_range"/>
    <arg name="min_add_scan_shift" default="0.1" doc="a moving distance of a map update[m]"/>
    <arg name="use_imu" default="true" doc="use imu or not"/>
    <arg name="imu_topic" default="imu_raw" />

    <arg name="lidar_frame" default="/rslidar" />
    <arg name="lidar_topic" default="/rslidar_points" />

    <!-- Static Transform Parameters -->
    <!-- 0.37 0 0.275   -1.57 -->
    <arg name="x" default="0" doc="static_transform_x lidar_to_base_link"/>
    <arg name="y" default="0.0" doc="static_transform_y lidar_to_base_link"/>
    <arg name="z" default="0" doc="static_transform_z lidar_to_base_link"/>
    <arg name="roll" default="0.0" doc="static_transform_roll lidar_to_base_link"/>
    <arg name="pitch" default="0.0" doc="static_transform_pitch lidar_to_base_link"/>
    <arg name="yaw" default="0" doc="static_transform_yaw lidar_to_base_link"/>

    <!-- Save Map -->
    <!-- <node pkg="pcl_ros" type="pointcloud_to_pcd" name="pointcloud_to_pcd" args="input:=/ndt_map _prefix:=$(find htc_ndt_mapping)/map/map_" output="screen"/> -->
    
    <!-- Generate Map -->
    <node pkg="ndt_mapping" type="htc_ndt_mapping" name="htc_ndt_mapping" output="screen">
      <param name="map_path" value="$(arg map_path)" />

      <param name="max_iter" value="$(arg max_iter)" />
      <param name="step_size" value="$(arg step_size)" />
      <param name="ndt_res" value="$(arg ndt_res)" />
      <param name="trans_eps" value="$(arg trans_eps)" />
      <param name="voxel_leaf_size" value="$(arg voxel_leaf_size)" />
      
      <param name="scan_rate" value="$(arg scan_rate)" />
      <param name="min_scan_range" value="$(arg min_scan_range)" />
      <param name="max_scan_range" value="$(arg max_scan_range)" />
      <param name="min_add_scan_shift" value="$(arg min_add_scan_shift)" />
      <param name="use_imu" value="$(arg use_imu)" />
      <param name="imu_topic" value="$(arg imu_topic)" />

      <param name="lidar_frame" value="$(arg lidar_frame)" />
      <param name="lidar_topic" value="$(arg lidar_topic)" />

      <param name="x" value="$(arg x)" />
      <param name="y" value="$(arg y)" />
      <param name="z" value="$(arg z)" />
      <param name="roll" value="$(arg roll)" />
      <param name="pitch" value="$(arg pitch)" />
      <param name="yaw" value="$(arg yaw)" />
    </node>

</launch>
